### ISSUE TITLE
relink_namespace in xml_node.c destroys documents

### ISSUE THREAD

**Issue by etm** [sparklemotion/nokogiri#1322]
File test1.xml:

``` xml
<a xmlns="uri:a" xmlns:b="uri:b">
  <b/>
</a>
```

File test2.xml

``` xml
<b xmlns="uri:b"/>
```

execute the following example:

``` ruby
require 'nokogiri'

a = Nokogiri::XML.parse(File.read('test1.xml'))
b = Nokogiri::XML.parse(File.read('test2.xml'))

a.root.xpath('/*/*[1]').each do |ele|
  ele.replace b.root
end

puts a.root.serialize(:encoding => 'UTF-8', :save_with => Nokogiri::XML::Node::SaveOptions::FORMAT | Nokogiri::XML::Node::SaveOptions::AS_XML)
```

This results in the following (which is very wrong):

``` text
<a xmlns="uri:a" xmlns:b="uri:b">
  <b/>
</a>
```

It should actually result in (both would be correct):

``` xml
<a xmlns="uri:a" xmlns:b="uri:b">
  <b:b/>
</a>
```

or

``` xml
<a xmlns="uri:a" xmlns:b="uri:b">
  <b xmlns="uri:b"/>
</a>
```

**relink_namespace in xml_node.c line 79ff seriously breaks things!**

A second example that shows that things are even worse (and not just an exception for standard namespaces):

File test1.xml:

``` xml
<a xmlns="uri:a" xmlns:c="uri:b">
  <b/>
  <c:b/>
</a>
```

File test2.xml:

``` xml
<b:b xmlns:b="uri:b"><x/></b>
```

Run with the ruby above and the result is:

``` xml
<a xmlns="uri:a" xmlns:c="uri:b">
  <b:b><x/></b:b>
  <c:b/>
</a> 
```

Thats very bad as well. Consider removing relink_namespace from functions such as replace. This way you have an excess of namespace declarations but at least the document is always correct.

To fix relink_namespace you have to rename nodes, whenever you remove namespaces (and the prefix of the parent namespace is different).

**Comment by etm** [comment]
A temporary (monkey patch) solution for replace is:

```ruby
module Nokogiri
  module XML
    class Node
      alias __replace replace
      def replace(node_or_tags)
        x = __replace(node_or_tags)
        x.xpath('.|.//*|.//@*').each do |rns|
          x.namespace_scopes.each do |nss|
            rns.namespace = nss if rns.namespace && rns.namespace.href == nss.href
          end
        end
        x                                                                                                                                                   
      end
    end
  end
end
```

As ***relink_namespace*** is used by other functions as well, it will be necessary to check if this solution is needed there as well (very possible). Should be fairly straightforward to port to c and directly include in  ***relink_namespace*** (*xml_node.c*).

**Comment by flavorjones** [comment]
This has been fixed since v1.6.8, I believe. Apologies for not responding before now.

Current behavior:

```ruby
#! /usr/bin/env ruby

require "nokogiri"

a = Nokogiri::XML.parse(<<~EOF)
<a xmlns="uri:a" xmlns:b="uri:b">
  <b/>
</a>
EOF

b = Nokogiri::XML.parse(<<~EOF)
<b xmlns="uri:b"/>
EOF

a.root.xpath('/*/*[1]').each do |ele|
  ele.replace b.root
end

a.root.serialize(:encoding => 'UTF-8', :save_with => Nokogiri::XML::Node::SaveOptions::FORMAT | Nokogiri::XML::Node::SaveOptions::AS_XML)
# => "<a xmlns=\"uri:a\" xmlns:b=\"uri:b\">\n" +
#    "  <b xmlns=\"uri:b\"/>\n" +
#    "</a>"


a = Nokogiri::XML.parse(<<~EOF)
<a xmlns="uri:a" xmlns:c="uri:b">
  <b/>
  <c:b/>
</a>
EOF

b = Nokogiri::XML.parse(<<~EOF)
<b:b xmlns:b="uri:b"><x/></b>
EOF

a.root.xpath('/*/*[1]').each do |ele|
  ele.replace b.root
end

a.root.serialize(:encoding => 'UTF-8', :save_with => Nokogiri::XML::Node::SaveOptions::FORMAT | Nokogiri::XML::Node::SaveOptions::AS_XML)
# => "<a xmlns=\"uri:a\" xmlns:c=\"uri:b\">\n" +
#    "  <b:b xmlns:b=\"uri:b\"><x/></b:b>\n" +
#    "  <c:b/>\n" +
#    "</a>"
```
